STACpopulator/extensions/base.py

```diff
@@ -0,0 +1,331 @@
+
+    @model_validator(mode="before")
+    @classmethod
+    def validate_jsonschema(cls, data):
+        """Validate the data model against the json schema, if given."""
+        # Load schema
+        uri =  cls._schema_uri.default
+        if uri is not None:
+            schema = json.load(open(uri))
+            validator_cls = jsonschema.validators.validator_for(schema)
+            validator_cls.check_schema(schema)
+            validator = validator_cls(schema)
+
+            attrs = {f"{cls._prefix.default}:{k}": v for (k,v) in data.items() if k not in cls._schema_exclude.default}
+            errors = list(validator.iter_errors(attrs))
+            if errors:
+                raise ValueError(errors)
+
+        return data
+
+    def apply(self, item, add_if_missing=True):
+        """Add extension for the properties of the dataset to the STAC item.
+        The extension class is created dynamically from the properties.
+        """
+        schema_uri = self.write_stac_schema() if self._schema_uri else None
+        ExtSubCls = metacls_extension(self._prefix, schema_uri=schema_uri)
+        item_ext = ExtSubCls.ext(item, add_if_missing=add_if_missing)
+        item_ext.apply(self.model_dump(mode="json", by_alias=True))
+        return item
+
+    def to_stac_schema(self) -> dict:
+        """Return the STAC schema for the extension."""
+        return {'type': 'object',
+                'required': ['type', 'properties'],
+                'properties': {'type': {'const': 'Feature'},
+                               'properties': {'$ref': str(self._schema_uri)}
+                               }
+                }
+    def write_stac_schema(self) -> str:
+        path = f"/tmp/{self._prefix}-schema.json"
+        with open(path, "w") as fh:
+            json.dump(self.to_stac_schema(), fh)
+        return path
+
+
+class BaseSTAC(BaseModel):
+    """Base class for STAC item data models.
+
+    Attributes
+    ----------
+    geometry : AnyGeometry
+        The geometry of the dataset.
+    bbox : list[float]
+        The bounding box of the dataset.
+    start_datetime : datetime
+        The start datetime of the dataset.
+    end_datetime : datetime
+        The end datetime of the dataset.
+    extensions : list[str]
+        Name of the class attributes that point to STAC extension helper classes. Those extension classes should have an `apply` method.
+    """
+    # STAC item properties
+    geometry: AnyGeometry | None
+    bbox: list[float]
+    start_datetime: datetime
+    end_datetime: datetime
+
+    model_config = ConfigDict(populate_by_name=True, extra="ignore", arbitrary_types_allowed=True)
+
+    # Helpers are automatically detected by being Helper subclasses
+    _helpers: list[str] = PrivateAttr([])
+
+    @property
+    def uid(self) -> str:
+        """Return a unique ID. When subclassing, use a combination of properties uniquely identifying a dataset."""
+        # TODO: Should this be an abstract method?
```

Comment:
Yeah I think it probably should.
Especially since the intention is for it to be overridden when this is subclassed.

Also this may cause problems if a subclass forgets to redefine this and then it is called multiple times within the subclass... it will give a different uid each time:

```python
>>> item = BaseSTAC(geometry=None, bbox=[], start_datetime='1999-01-01', end_datetime='1999-01-02')
>>> item.uid
'd33916d8-7bf6-4bf3-9659-e590badc96c6'
>>> item.uid
'b6729ec6-ced6-4076-99fe-e0df321bd163'
```